inference_worker.py

```diff
@@ -0,0 +1,210 @@
+            channel = grpc.insecure_channel(FLAGS.address)
+            stub = inference_service_pb2_grpc.InferenceServiceStub(channel)
+            return stub.GetConfig(inference_service_pb2.GetConfigRequest())
+        except Exception:  # pylint: disable=broad-except
+            print("Waiting for server")
+            time.sleep(1)
+
+
+def wrapped_model_inference_fn(config):
+    """Wraps dual_net.model_inference_fn in a loop & RPC ops.
+
+    The loop runs forever: the top of the loop issues a GetFeatures RPC to
+    fetch input features, the bottom of the loop isses a PutOutputs RPC to
+    write the inference output back to the server.
+
+    Args:
+        config: InferenceWorkerConfig.
+
+    Returns:
+        A tensor op that drives the model's infinite loop.
+    """
+
+    value_output_size = config.batch_size
+    policy_output_size = config.batch_size * (go.N * go.N + 1)
+
+    def loop_condition(a, unused_b):
+        """Loop condition for the tf.while_loop op.
+
+        Args:
+            a: a constant 0
+            unused_b: a string placeholder (to satisfy the requirement that a
+                      while_loop's condition and body accept the same args as
+                      the loop returns).
+
+        Returns:
+            A TensorFlow subgraph that returns true.
+        """
+
+        # TensorFlow will reject a loop unless its condition contains at least
+        # one comparison. So we use (a < 1), initialize a to 0, and never
+        # increment it.
+        return tf.less(a, 1)
+
+    def loop_body(a, unused_b):
+        """Loop body for the tf.while_loop op.
+
+        Args:
+            a: a constant 0
+            unused_b: a string placeholder (to satisfy the requirement that a
+                      while_loop's condition and body accept the same args as
+                      the loop returns).
+
+        Returns:
+            A TensorFlow subgraph.
+        """
+
+        # Request features features.
+        raw_response = tf.contrib.rpc.rpc(
+            address=config.address,
+            method=config.get_features_method,
+            request="",
+            protocol="grpc",
+            fail_fast=True,
+            timeout_in_ms=0,
+            name="get_features")
+
+        # Decode features from a proto to a flat tensor.
+        _, (batch_id, flat_features) = decode_proto_op.decode_proto(
+            bytes=raw_response,
+            message_type='minigo.GetFeaturesResponse',
+            field_names=['batch_id', 'features'],
+            output_types=[dtypes.int32, dtypes.float32],
+            descriptor_source=config.descriptor_path,
+            name="decode_raw_features")
+
+        # Reshape flat features.
+        features = tf.reshape(
+            flat_features, [-1, go.N, go.N, features_lib.NEW_FEATURES_PLANES],
+            name="unflatten_features")
+
+        # Run inference.
+        policy_output, value_output, _ = dual_net.model_inference_fn(
```

Comment:
if i understand this right, this means that the model checkpoints loaded now need to match the hyperparams defined here -- i.e., we've lost the ability to load arbitrary depth/width nets that had been frozen as .pbs, correct?

Reply:
Sadly yes, if you want to run with the InferenceServer. At least until I figure out how we can build C++ TensorFlow with the RPC & proto ops linked in.